components/ntp_background_images/browser/ntp_background_images_service.cc

```diff
@@ -6,82 +6,406 @@
-  };
-  demo->logo_alt_text = "Technikke: For music lovers.";
-  demo->logo_company_name = "Technikke";
-  demo->logo_destination_url = "https://brave.com";
-  return demo.get();
+bool IsValidSuperReferrerComponentInfo(const base::Value& component_info) {
+  if (!component_info.is_dict())
+    return false;
+
+  if (!component_info.FindStringKey(kPublicKey))
+    return false;
+  if (!component_info.FindStringKey(kComponentID))
+    return false;
+  if (!component_info.FindStringKey(kCompanyName))
+    return false;
+
+  return true;
+}
+
+const net::NetworkTrafficAnnotationTag& GetNetworkTrafficAnnotationTag() {
+  static const net::NetworkTrafficAnnotationTag network_traffic_annotation_tag =
+      net::DefineNetworkTrafficAnnotation("ntp_background_images_service", R"(
+        semantics {
+          sender:
+            "Brave NTP Background Images Service"
+          description:
+            "Download Super referrer component mapping table"
+          trigger:
+            "When the browser needs to check whether this installation was by
+             a 'super referrer'"
+          data: "component mapping table"
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          policy_exception_justification:
+            "Not implemented."
+        })");
+  return network_traffic_annotation_tag;
 }
 
 }  // namespace
 
 NTPBackgroundImagesService::NTPBackgroundImagesService(
-    component_updater::ComponentUpdateService* cus)
-    : weak_factory_(this) {
-  // Flag override for testing or demo purposes
-  base::FilePath forced_local_path(
-      base::CommandLine::ForCurrentProcess()->GetSwitchValueNative(
-          switches::kNTPBrandedDataPathForTesting));
-  if (!forced_local_path.empty()) {
-    LOG(ERROR)
-        << "NTP Sponsored Image package will be loaded from local path at: "
-        << forced_local_path.LossyDisplayName();
-    OnComponentReady(forced_local_path);
+    component_updater::ComponentUpdateService* cus,
+    PrefService* local_pref,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : component_update_service_(cus),
+      local_pref_(local_pref),
+      url_loader_factory_(std::move(url_loader_factory)),
+      weak_factory_(this) {
+}
+
+NTPBackgroundImagesService::~NTPBackgroundImagesService() = default;
+
+void NTPBackgroundImagesService::Init() {
+  if (LoadTestData())
+    return;
+
+#if BUILDFLAG(ENABLE_BRAVE_REFERRALS)
+  if (base::FeatureList::IsEnabled(features::kBraveNTPSuperReferrerWallpaper)) {
+    // Download mapping table first. Then determine which component should be
+    // registered.
+    DownloadSuperReferrerMappingTable();
+    return;
+  }
+#endif
+
+  StartSponsoredImagesComponent();
+}
+
+void NTPBackgroundImagesService::DownloadSuperReferrerMappingTable() {
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = GURL(kSuperReferrerMappingTableURL);
```

Comment:
I don't understand why we're making a separate download for something here, especially when it only happens on startup. That means there is no defined time when it will be updated (I rarely restart my browser). If we need information to determine which component to download, that information should also be in a component that we always download, not a separate one-off download that has no defined schedule to run.

Reply:
Mapping table doesn't need to be updated in runtime.
If currently running browser is super referrer and its campaign ends, super referrer component will be updated. If SR component is empty, that means its campaign is ended. Then, this service will register sponsored images component.